### PR TITLE
On-the-fly resample of binned data.

### LINKED ISSUE #1397: Memory use when plotting binned data
#1387 automatically histograms binned data when a plot is created. However, in practice this is often not feasible: 
- Instruments with high resolution and high pixel count, e.g., a diffractometer with 1 million pixels and 10000 TOF bins (bins for plotting, not in the binned data).
- 3D or 4D cases such as `(Qx, Qy, Qz)` for single-crystal data or `(Qx, Qy, Qz, delta-E)` for spectrometers.

In both cases memory requirements may prevent direct histogramming of the full volume at full resolution. I think we therefore need to handle this just like the resampling step that is undertaken for display purposes: Histogram directly into the edges that we need for display. This also involves slicing before histogramming, which is one of the key advantages provided by binned data.

For the implementation, maybe the design is best supported by two distinct models, a `HistogramModel` (the existing one) and the `BinnedModel`?

### PR DESCRIPTION
**Pull request by SimonHeybrock**
Fixes #1397.

Instead of histogramming directly, which may use more memory than available, we bin into image pixel (as determined by resolution). This still requires a temporary copy of all the events at this point. Future updates to this solution could reduce this requirements.

The overall approach is the extract the resampling code into a new "model", `ResamplingModel`, with 3 implementations:
- Resampling binned data.
- Resampling histogrammed (counts) data.
- Resampling non-counts dense data.

Reviewers:
- Try loading some event data and plot it, `events.plot()`. Try to zoom in, pan, ....
- Compare to `events.bins.sum().plot()` or sc.histogram(events.bins, tof_edges).
- You can try the following as a 3D example (also try the profile view here):
```python
import numpy as np
import scipp as sc
from scipp.plot import plot
L = 4
M = 5
N = 5
s = 100000
x = 10.0 * np.arange(L)
y = 10.0 * np.arange(M)
z = 10.0 * np.arange(N)
xmesh, ymesh, zmesh = np.meshgrid(y, x, z)

distance = 1.0
xpos = []
ypos = []
zpos = []
for i in range(L):
    for j in range(M):
        for k in range(N):
            xpos.append(np.random.normal(xmesh[i, j, k], distance, size=(s,)))
            ypos.append(np.random.normal(ymesh[i, j, k], distance, size=(s,)))
            zpos.append(np.random.normal(zmesh[i, j, k], distance, size=(s,)))
events = sc.DataArray(
    data=sc.Variable(dims=['event'], values=np.ones(L*N*M*s)),
    coords={
        'x':sc.Variable(dims=['event'], values=np.array(xpos).ravel()),
        'y':sc.Variable(dims=['event'], values=np.array(ypos).ravel()),
        'z':sc.Variable(dims=['event'], values=np.array(zpos).ravel()),
           })

xbins = sc.Variable(dims=['x'], values=np.linspace(-10,50, num=40))
ybins = sc.Variable(dims=['y'], values=np.linspace(0,50, num=40))
zbins = sc.Variable(dims=['z'], values=np.linspace(-10,50, num=40))
binned = sc.bin(events, edges=[zbins, ybins, xbins])
binned.masks['x'] = xbins['x',1:] < 0.0*sc.units.one
plot(binned, resolution=40, vmin=0, vmax=10000)
``` 

Other changes:
- `dataset::bin` has been refactored so custom coord dicts can be provided. This allows for rebinning to custom edges for a variable, without the need to construct a data array first.
- Label-based slicing refactored so custom coord dicts can be provided. We can thus just return the slice params (`get_slice_params`), to slice a variable instead of a data array.
- Label-based slicing with a range and a bin-edge coord now includes the upper bin. If we first slice and then rebin with the same bounds we do now *not* suffer from an empty final bin.
- update `nanmin` and friends to support integers.

Future work:
- #1469 to either replace hack by better mechanism, or remove the coord conversion.

### CODE REVIEW

**Comment by SimonHeybrock** [comment]
Sorry for requesting reviews by so many people. Note necessary to get code review from all of you, but it would be good if you could at least try the new feature, since it will be a central part of UX when working with event data.

**Comment by jl-wynen** [comment]
Is it supposed to look like this?
Screenshot_20201214_185851

**Comment by SimonHeybrock** [comment]
> Is it supposed to look like this?
> Screenshot_20201214_185851

Good catch, apparently masks are not getting transposed.

**Comment by nvaytet** [comment]
@jl-wynen can you try the example with masks here and see what it looks like? https://scipp.github.io/visualization/plotting-2d-data.html#Plotting-masks

**Comment by SimonHeybrock** [comment]
> @jl-wynen can you try the example with masks here and see what it looks like? https://scipp.github.io/visualization/plotting-2d-data.html#Plotting-masks

Make sure to update your branch, since I pushed a fix earlier.

**Comment by nvaytet** [comment]
I couldn't see that Jan Lukas had pressed the transpose button. Should this also be a toggle button where the color changes if you pressed it? Not sure if it makes sense in all cases..

**Comment by SimonHeybrock** [comment]
> I couldn't see that Jan Lukas had pressed the transpose button. Should this also be a toggle button where the color changes if you pressed it? Not sure if it makes sense in all cases..

I think it should be a toggle button: It expresses the difference to what was given on command line (either natural order, or explicit axis order).

**Comment by jl-wynen** [comment]
I didn't press any button. I only moved the slider.

**Comment by jl-wynen** [comment]
Also, I just pulled and the plot still looks the same.

**Comment by jl-wynen** [comment]
> @jl-wynen can you try the example with masks here and see what it looks like? https://scipp.github.io/visualization/plotting-2d-data.html#Plotting-masks

Looks good. But I had to change the unit of `d3` because I got an error (unrelated to this PR):
``` 
RuntimeError: Unsupported unit as result of multiplication: (K) * (m)
```

**Comment by nvaytet** [comment]
Hmm maybe it is related to this PR?
Since this is working in master... I'm guessing there is some normalization by coord bin width that could be going on?
This would give the `K` of the data * the `m` of the coordinates.
Previously, I solved this by making a copy of the data and setting the units to dimensionless.

**Comment by nvaytet** [comment]
I also get the units issue with the example from the docs.
However, the example given above works well for me
Screenshot at 2020-12-15 09-37-46

**Comment by jl-wynen** [comment]
You're right. The blobs are placed correctly on my machine, too. I was only looking at the purple-black background which looks like a glitch. But it seems to be a feature of the data.

Also, clicking any of the x, y, z buttons produces a correct figure.

**Comment by SimonHeybrock** [comment]
> You're right. The blobs are placed correctly on my machine, too. I was only looking at the purple-black background which looks like a glitch. But it seems to be a feature of the data.
> 
> Also, clicking any of the x, y, z buttons produces a correct figure.

Do you see a *discontinuity* in the blobs at the transition to the gray-scale section, as in your original screenshot? Or is that fixed now?

File: python/src/scipp/plot/resampling_model.py
Comment (nvaytet) [inline python/src/scipp/plot/resampling_model.py:124]: Is this only applicable to `spectrum` or is it a fix for neutron data?
If it is only for spectrum, then you could have `if "spectrum" in self._array.meta:`.
If it isn't just for spectrum, then something like `if not bin_edge`:`

File: python/src/scipp/plot/resampling_model.py
Comment (nvaytet) [inline python/src/scipp/plot/resampling_model.py:163]: Is `astype` here always making a copy? Isn't that potentially expensive?

**Review by nvaytet — CHANGES_REQUESTED** [review]
A few more things:

- profile doesn't seem to work with `projection="1d"`
- Plotting binned data in 1d is not supported? e.g. `plot(binned['x', 0]['y', 0])`
- 2d plots for dense data is broken for non-counts units when coords also have units, e.g the example in the docs (2d plotting)

For this last point, this is not being caught by the tests, so I suggest adding units to the coordinates in the `make_dense_dataset` inside the `plot_helper.py` file. Something like replacing `dim_list` with
```Py
dims_and_units = {
    'tof': sc.units.us,
    'x': sc.units.m,
    'y': sc.units.one,
    'z': sc.units.m,
    'Q_x': sc.units.angstrom
}
dim_list = list(dims_and_units.keys())
```
and giving units to coords with `unit=dims_and_units[dim]`.

File: dataset/test/slice_by_value_test.cpp
Comment (OwenArnold) [inline dataset/test/slice_by_value_test.cpp:191]: This looks correct.

**Review by OwenArnold — COMMENTED** [review]
A few comments from some hands-on and a quick look at slicing.

- Resampling behaviour is awesome.
-  Have you stress-tested this with larger event files? My concern is that the resampling could be too expensive and break interactivity or lead to a crash?

I'll take a closer look at the code tomorrow.

File: python/src/scipp/plot/resampling_model.py
Comment (SimonHeybrock) [reply python/src/scipp/plot/resampling_model.py:163]: See our discussion from 10 days ago: We *have* to make a copy because we have to rescale before rebin. Since this is only used for non-count data we currently to not have many performance-critical applications for this, I believe.

**Comment by SimonHeybrock** [comment]
>     * Have you stress-tested this with larger event files? My concern is that the resampling could be too expensive and break interactivity or lead to a crash?

Largest I tried was a 6.5 GB WISH file. It took 6 seconds for the initial plot to show up.

**Comment by SimonHeybrock** [comment]
> > @jl-wynen can you try the example with masks here and see what it looks like? https://scipp.github.io/visualization/plotting-2d-data.html#Plotting-masks
> 
> Looks good. But I had to change the unit of `d3` because I got an error (unrelated to this PR):
> 
> ```
> RuntimeError: Unsupported unit as result of multiplication: (K) * (m)
> ```

Yes, I broke this, fixed in an update which I will push soon.

File: python/src/scipp/plot/resampling_model.py
Comment (nvaytet) [reply python/src/scipp/plot/resampling_model.py:164]: yes but in master, we are making a copy of at most a 2D array, because the copy was made after having sliced away the additional dimensions. Here, i think you are copying the entire ND data, right?

Comment (SimonHeybrock) [reply python/src/scipp/plot/resampling_model.py:164]: True, but this would disappear once we support rebinning non-counts data in one way or another.

**Comment by SimonHeybrock** [comment]
> * profile doesn't seem to work with `projection="1d"`

Please try again now.

> * Plotting binned data in 1d is not supported? e.g. `plot(binned['x', 0]['y', 0])`

Yes, I think I would propose to implement this later. We don't really have a pressing application for this right now.

**Comment by nvaytet** [comment]
> Yes, I think I would propose to implement this later. We don't really have a pressing application for this right now.

Can we exit gracefully instead of getting something like
```Py
~/code/scipp/jupyter/scipp/plot/tools.py in mask_to_float(mask, var)
    122     Return an array of masks as floats.
    123     """
--> 124     return np.where(mask, var, None).astype(np.float32)
    125 
    126 

ValueError: setting an array element with a sequence.
```

**Review by nvaytet — APPROVED** [review]
Looks good to me, but I would say we need at least one more approval before we merge

File: variable/slice.cpp
Comment (OwenArnold) [inline variable/slice.cpp:90]: This applies the new behaviour - might be worth explaining this here?

**Review by OwenArnold — CHANGES_REQUESTED** [review]
Looks good. Main change is for the slicing documentation.

File: core/include/scipp/core/dimensions.h
Comment (SimonHeybrock) [reply core/include/scipp/core/dimensions.h:38]: This constructor came from the need of writting common code for data array and dataset. The latter gives and unordered map as dims so order is undefined, but at least we can then use common methods that do not depend on order.

Whether `add` or `addInner` is used does not matter here.

File: docs/user-guide/slicing.ipynb
Comment (nvaytet) [inline docs/user-guide/slicing.ipynb:440]:
```suggestion
    "If an interval is provided when slicing with a bin-edge coordinate, the range of bins *containing* the interval bounds (*including* the left as well as the right bin) is selected:"
```

**Review by nvaytet — CHANGES_REQUESTED** [review]

**Review by nvaytet — APPROVED** [review]

**Pull request merged**
